osquery/sql/sqlite_collation.cpp

```diff
@@ -0,0 +1,257 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <string>
+
+#include <sqlite3.h>
+
+namespace osquery {
+/**
+ * @brief Compares the position of epoch delimiter between two versions.
+ * Return difference of epoch segment length if both versions have epoch.
+ * Return 1 if left version has epoch but not the right.
+ * Return -1 if right version has epoch but not the left.
+ * Return 0 if epoch doesn't exist in either version.
+ */
+static int compareEpoch(int lLen,
+                        const char* lVer,
+                        int rLen,
+                        const char* rVer) {
+  auto lEpoch = strcspn(lVer, ":");
+  auto rEpoch = strcspn(rVer, ":");
+
+  if (lEpoch != lLen && rEpoch != rLen) {
+    return lEpoch - rEpoch;
```

Comment:
this is comparing the ascii values? I think I have w questions...

First, the comment says this returns -1,0,1. But I think this will return any int. 

Second, does this correctly handle 2 character epochs?

Reply:
Not at this step. It's comparing match position. If both strings have epoch and say at pos 2, then it returns 0 and will fall into the loop where the actual values preceding the epoch gets compared. If only one string has an epoch, or if an epoch has more preceding characters, then this will return a less than or greater than.